Increase robustness of LwsApiCall implementation

vikramdattu opened the pull request:
## What changed
Improved robustness of the LWS (libwebsockets) API call implementation for signaling.

## Why
The existing implementation had several reliability issues:
- Partial HTTP body writes were rescheduled but the buffer pointer/size weren't updated, causing duplicate data
- WSS message sends had no retry mechanism — a single CVAR_WAIT timeout would fail the entire send
- No ping/pong keepalive for the WSS connection
- Peer-initiated close always returned error even for normal closure
- No tracking of in-progress message reception, leading to premature termination during multi-fragment receives

## How
- **HTTP partial write fix**: Update `body` pointer and `bodySize` on partial write before rescheduling
- **WSS ping/pong**: Add `LWS_CALLBACK_TIMER` handler for periodic keepalive, schedule via `lws_set_timer_usecs`
- **Close callback**: Distinguish normal vs abnormal closure using `LWS_CLOSE_STATUS_NORMAL` and `receiveMessage` state
- **Receive callback**: Track multi-fragment message reception with `receiveMessage` atomic; reset buffer on final fragment; check shutdown before scheduling writes
- **Writeable callback**: Add send pipe choke detection via `lws_send_pipe_choked`; handle partial writes with `MEMMOVE`; signal `messageResult` on success
- **Send retry logic**: Add `MAX_LWS_SEND_RETRY_COUNT` (3) retries in `writeLwsData` with proper CVAR_WAIT timeout handling
- **HTTP/1.1 forcing**: Force HTTP/1.1 via ALPN to reduce memory footprint on constrained platforms (ESP32)

## Review comments addressed
- Replaced `CHK(FALSE, retValue)` with `STATUS_SIGNALING_LWS_CALL_FAILED`
- Fixed CVAR_WAIT timeout bypassing retry count (store result, reset status, continue loop)
- Added `DLOGD` for retry count on CVAR_WAIT wake-up
- Moved `MAX_RETRY_COUNT` to `#define MAX_LWS_SEND_RETRY_COUNT` in LwsApiCalls.h
- Added comment explaining H2 quirk flag + HTTP/1.1 rationale

## Testing done
- 277/278 non-integration tests pass (1 pre-existing sandbox socket failure)
- Clang-format compliance check passes
- Rebased onto latest `develop` branch

vikramdattu committed:
Increase robustness of LwsApiCall implementation

 - Handle partial writes by sending data in multiple iterations
 - Use retries when message send fails
 - Track and handle message receive if in parts using `receiveMessage` var

sirknightj commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -569,8 +607,10 @@ STATUS lwsCompleteSync(PLwsCallInfo pCallInfo)
+    connectInfo.ssl_connection = LCCSCF_USE_SSL | LCCSCF_H2_QUIRK_OVERFLOWS_TXCR; // Add flag to handle H2 flow control
     connectInfo.port = SIGNALING_DEFAULT_SSL_PORT;
+    connectInfo.alpn = "http/1.1";     // Force HTTP/1.1 only
+    connectInfo.protocol = "http/1.1"; // Force HTTP/1.1 protocol
```
Wondering the purpose of the H2 flag if 1.1 is forced, is it for latency/performance optimization?

It seems that HTTP 2 should from this quick test:
```
curl -sI https://m-xxxxxxxx.kinesisvideo.us-west-2.amazonaws.com -o/dev/null -w '%{http_version}\n'
2
curl -sI https://v-xxxxxxxx.kinesisvideo.us-west-2.amazonaws.com -o/dev/null -w '%{http_version}\n'
2
```

vikramdattu replied:
I added this to make this ws connection work more reliably on ESP platforms. Helps reduce memory consumption a bit and works much reliably. BTW, do we really gain much with http2 for our use case?

Now that, I have lot of optimisations in place, I can test http2 with ESP32 and revert the change.

sirknightj commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -1907,6 +1948,9 @@ STATUS writeLwsData(PSignalingClient pSignalingClient, BOOL awaitForResponse)
     SERVICE_CALL_RESULT result;
 
+    UINT32 retryCount = 0;
+    const UINT32 MAX_RETRY_COUNT = 3;
```
Thinking this should be a `#define` (very very minor memory reduction)

sirknightj reviewed (commented):
These are good improvements..

Wanted to know if you had a method or setup to test these failure recovery scenarios and if we could add something for it? Seems like it would be a lot of additional effort.

vikramdattu commented on src/source/Signaling/LwsApiCalls.c:
```diff
@@ -1920,22 +1964,36 @@ STATUS writeLwsData(PSignalingClient pSignalingClient, BOOL awaitForResponse)
 
         if (offset != size && result == SERVICE_CALL_RESULT_NOT_SET) {
             CHK_STATUS(CVAR_WAIT(pSignalingClient->sendCvar, pSignalingClient->sendLock, SIGNALING_SEND_TIMEOUT));
+            retryCount++;
```
@sirknightj looks a bit odd, but here is what's happening:
- If `CVAR_WAIT` returns on timeout, we exit anyway.
- Only if it was returned early, we increment the `retryCount` and go again for wait if `offset != size`.
 else we stop iterating and go for the rest of the code path.
 
So, retryCount is basically, making it `retry` if wakeup was within timeout.
Let me know if you have a better suggestion to handle `sendCvar` wake-ups without just giving up entirely.

vikramdattu committed:
Address review comments for LwsApiCalls robustness

- Move MAX_RETRY_COUNT to #define MAX_LWS_SEND_RETRY_COUNT in header
- Fix CVAR_WAIT timeout bypassing retry count: store result, reset status
  on timeout, increment retryCount and continue loop instead of goto
- Add DLOGD for retry count on CVAR_WAIT wake-up
- Replace CHK(FALSE, retValue) with STATUS_SIGNALING_LWS_CALL_FAILED
- Add comment explaining H2 quirk flag + HTTP/1.1 forcing rationale
